### PR TITLE
Fix restoring bug and add more test cases

### PR DESCRIPTION
**Pull request by OsmondJiang**

### CODE REVIEW

File: src/docfx/restore/Restore.cs [hunk 1]
```diff
@@ -57,6 +57,7 @@ public static (string restoreDir, string url, string refSpec) GetGitRestoreInfo(
         private static async Task<string> FetchOrCloneDependentRepo(string href)
         {
             var (restoreDir, url, rev) = GetGitRestoreInfo(href);
+            var lockPath = Path.Combine(restoreDir.Replace("/git/", "/lock/", StringComparison.OrdinalIgnoreCase), ".lock");
```
Comment (yufeih) [inline src/docfx/restore/Restore.cs]: Should not use "Replace"

File: src/docfx/restore/Restore.cs [hunk 2]
```diff
@@ -34,7 +34,7 @@ public static (string restoreDir, string url, string refSpec) GetGitRestoreInfo(
             var refSpec = (string.IsNullOrEmpty(uri.Fragment) || uri.Fragment.Length <= 1) ? "master" : uri.Fragment.Substring(1);
             var url = uri.GetLeftPart(UriPartial.Path);
             var repo = Path.Combine(uri.Host, uri.AbsolutePath.Substring(1));
-            var dir = Path.Combine(s_restoreDir, repo);
+            var dir = Path.Combine(s_restoreDir, repo, refSpec);
```
Comment (superyyrrzz) [inline src/docfx/restore/Restore.cs]: >refSpec [](start = 55, length = 7)

not safe? a valid branch name may contain invalid path chars.

File: test/docfx.Test/common/TestHelper.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            }
+
+            return docsets;
+        }
+
+        public static void VerifyJsonContainEquals(JToken expected, JToken actual)
+        {
+            if (expected is JArray expectedArray)
+            {
+                if (actual is JArray actualArray)
```
Comment (superyyrrzz) [inline test/docfx.Test/common/TestHelper.cs]: >if (actual is JArray actualArray) [](start = 16, length = 33)

directly Assert.True(actual is JArray)?

**Comment by superyyrrzz** [comment]
>                     restoreChild(childHref);

check if child repo is dependent on parent repo to avoid dead loop?

---
Refers to: src/docfx/restore/Restore.cs:51 in 2a996ca. [](commit_id = 2a996ca685e1c97ed4c17ff13aa222000ba663f1, deletion_comment = False)

File: test/docfx.Test/common/TestHelper.cs (hunk 3 above)
Comment (vwxyzh) [inline test/docfx.Test/common/TestHelper.cs]: Assert.IsType<JArray>(actual);

File: test/docfx.Test/common/TestHelper.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+                    }
+                }
+                else
+                {
+                    Assert.True(actual is JArray, $"Array expected: {actual}");
+                }
+            }
+            else if (expected is JObject expectedObject)
+            {
+                if (actual is JObject actualObject)
```
Comment (vwxyzh) [inline test/docfx.Test/common/TestHelper.cs]: `Assert.IsType<JObject >(actual);`

File: test/docfx.Test/common/TestHelper.cs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+                    }
+                }
+                else
+                {
+                    Assert.True(actual is JArray, $"Object expected: {actual}");
+                }
+            }
+            else
+            {
+                Assert.Equal(((JValue)expected).Value, ((JValue)actual).Value);
```
Comment (vwxyzh) [inline test/docfx.Test/common/TestHelper.cs]: `Assert.Equal(expected, actual);`

File: test/docfx.Test/lib/ProcessUtilityTest.cs [hunk 6]
```diff
@@ -25,7 +25,7 @@ public static class ProcessUtilityTest
 
                 File.Delete(fileName);
                 return Task.FromResult(0);
-            }, Path.GetFullPath($"{fileName}.lock"))));
+            }, $"{fileName}.lock")));
```
Comment (vwxyzh) [inline test/docfx.Test/lib/ProcessUtilityTest.cs]: What hell of indent!!

File: test/docfx.Test/restore/RestoreTest.cs [hunk 7]
```diff
@@ -3,22 +3,34 @@
...
+using Newtonsoft.Json.Linq;
 using Xunit;
 
 namespace Microsoft.Docs.Test
 {
     public static class RestoreTest
     {
+        public static readonly TheoryData<string> Specs = new TheoryData<string>();
+
+        static RestoreTest()
```
Comment (vwxyzh) [inline test/docfx.Test/restore/RestoreTest.cs]: avoid type initializer

File: stylecop.ruleset [hunk 8]
```diff
@@ -51,6 +51,8 @@
     <Rule Id="CA1063" Action="Warning" />
     <Rule Id="CA1065" Action="Warning" />
     <Rule Id="CA1301" Action="Warning" />
+    <!-- CA1307: Specify StringComparison -->
+    <Rule Id="CA1307" Action="None" />
```
Comment (vwxyzh) [inline stylecop.ruleset]: Warning is make sense.

File: test/docfx.Test/lib/ProcessUtilityTest.cs (hunk 6 above)
Comment (OsmondJiang) [reply test/docfx.Test/lib/ProcessUtilityTest.cs]: the stylecop doesn't give me any error or warnings, so i believe it's valid?

---
In reply to: [183932244](https://github.com/dotnet/docfx/pull/2666#discussion_r183932244) [](ancestors = 183932244)

File: test/docfx.Test/common/TestHelper.cs [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+            return docsets;
+        }
+
+        public static void VerifyJsonContainEquals(JToken expected, JToken actual)
+        {
+            if (expected is JArray expectedArray)
+            {
+                Assert.IsType<JArray>(actual);
+                if (actual is JArray actualArray)
```
Comment (vwxyzh) [inline test/docfx.Test/common/TestHelper.cs]: After `Assert.IsType`, `is` should be direct cast `actualArray = (JArray)actual;`

File: test/docfx.Test/lib/ProcessUtilityTest.cs (hunk 6 above)
Comment (yufeih) [reply test/docfx.Test/lib/ProcessUtilityTest.cs]: stylecop is disable in test, too many false positives. you can reorder to the parameter to make it *look good*: `Lock(path, action)`

**Pull request merged**